### PR TITLE
[REVIEW] Add MultiIndex support for DataFrame join

### PR DESCRIPTION
**Pull request by rjzamora**
Adds support for `MultiIndex`-based dataframes within the `DataFrame.join` Python API.  A new test has been added to `python/cudf/cudf/tests/test_joining.py` (`test_join_multi`) -- It considers join operations with integer, string, and categorical `MultiIndex` components.

### CODE REVIEW

**Comment by rjzamora** [comment]
**Note**:  For `MultiIndex`-based frames, this implementation requires the left and right indices to have the same list of column names (same number of columns, with all names overlapping).  Although a bit more logic can be added to handle the more-general case (partial overlap, etc), I would prefer to have that functionality added in a seperate PR (at a later time).  If there are objections to this approach (the general-case functionality has an immediate motivation/need/use-case), I can certainly spend more time here.

The motivation for the existing functionality in this PR is to unblock `split_out` support (#2994 )

File: python/cudf/cudf/tests/test_joining.py [hunk 1]
```diff
@@ -735,3 +735,56 @@ def test_join_empty_table_dtype():
     pd_merge = left.merge(right, how="left", left_on=["a"], right_on=["b"])
     gd_merge = gleft.merge(gright, how="left", left_on=["a"], right_on=["b"])
     assert_eq(pd_merge["a"].dtype, gd_merge["a"].dtype)
+
+
+@pytest.mark.parametrize("how", ["outer", "inner", "left", "right"])
+@pytest.mark.parametrize("cat", [True, False])
+def test_join_multi(how, cat):
+
+    size = 10
```
Comment (cwharris) [inline python/cudf/cudf/tests/test_joining.py]: I'm uncomfortable with this small of a size, as we're using it to produce values for exact-match, and it has a pretty good likelihood to not produce a given value. Actually, I'd prefer at least part of the test data to be declared explicitly for the same reason. If I'm not mistaken, the odds of `np.arange(size)` producing overlapping results for both cudf and pandas dataframes is extremely low, even for high-ish values of `size`. tldr; I'd like to see more-deterministic tests.

File: python/cudf/cudf/tests/test_joining.py [hunk 2]
```diff
@@ -735,3 +735,56 @@ def test_join_empty_table_dtype():
...
+
+    # Make sure columns are in the same order
+    columns = pdf_result.columns.values
+    gdf_result = gdf_result[columns]
+    pdf_result = pdf_result[columns]
+    if how != "outer":
+        pdf_result.index.name = "bob"
+        gdf_result.index.name = "mary"
+        pd.util.testing.assert_frame_equal(
+            gdf_result.to_pandas().reset_index(drop=True),
```
Comment (cwharris) [inline python/cudf/cudf/tests/test_joining.py]: Consider using `cudf.testing.assert_eq(...)`, possibly with `compare_dtype=False` argument. Not sure if that's identical, or even helpful. Just throwing it out there.

File: python/cudf/cudf/tests/test_joining.py [hunk 3]
```diff
@@ -735,3 +735,56 @@ def test_join_empty_table_dtype():
...
+            pdf_result.reset_index(drop=True),
+        )
+    else:
+        _check_series(pdf_result["a1"].fillna(-1), gdf_result["a1"].fillna(-1))
```
Comment (cwharris) [inline python/cudf/cudf/tests/test_joining.py]: There's no null inputs for this test, so I'm assuming the `fillna`'s are for non-inner join results?

File: python/cudf/cudf/tests/test_joining.py [hunk 4]
```diff
@@ -735,3 +735,56 @@ def test_join_empty_table_dtype():
...
+        {
+            "a1": np.arange(size),
+            "b": np.random.randint(0, 2, size=size),
+            "c": np.random.choice(["dog", "cat"], size=size),
+        }
+    )
+    if cat:
+        df1["c"] = df1["c"].astype("category")
+    df1 = df1.set_index(index)
+    gdf1 = cudf.from_pandas(df1)
```
Comment (cwharris) [inline python/cudf/cudf/tests/test_joining.py]: `cudf.from_pandas` forwards to `cudf.Dataframe.from_pandas`, which has `nans_to_nulls=True`. If you include null-value tests, that might need to be accounted for.

File: python/cudf/cudf/tests/test_joining.py [hunk 5]
```diff
@@ -735,3 +735,56 @@ def test_join_empty_table_dtype():
...
+            "b": np.random.randint(0, 4, size=size),
+            "c": np.random.choice(["bird"], size=size),
+        }
+    )
+    if cat:
+        df2["c"] = df2["c"].astype("category")
+    df2 = df2.set_index(index)
+    gdf2 = cudf.from_pandas(df2)
+
+    gdf_result = gdf1.join(gdf2, how=how, sort=True).sort_values(["a1", "a2"])
```
Comment (cwharris) [inline python/cudf/cudf/tests/test_joining.py]: What's the difference between `sort=True` and `sort_values(...)`?

File: python/cudf/cudf/core/dataframe.py [hunk 6]
```diff
@@ -2449,17 +2449,47 @@ def join(
         rhs = DataFrame()
 
         # Creating unique column name to use libgdf join
-        idx_col_name = str(random.randint(2 ** 29, 2 ** 31))
+        def _unique_tmp_name(l, r, names):
```
Comment (cwharris) [inline python/cudf/cudf/core/dataframe.py]: Why not use `uuid.uuid4` and forget the loop?

File: python/cudf/cudf/tests/test_joining.py (hunk 1 above)
Comment (rjzamora) [reply python/cudf/cudf/tests/test_joining.py]: > ...the odds of np.arange(size) producing overlapping results for both cudf and pandas dataframes is extremely low...

I agree that more-deterministic tests would be ideal.  Hoever, I'm not really sure what you mean by this.  either way - I'll try to make improvements here.

File: python/cudf/cudf/tests/test_joining.py [hunk 7]
```diff
@@ -735,3 +735,56 @@ def test_join_empty_table_dtype():
...
     assert_eq(pd_merge["a"].dtype, gd_merge["a"].dtype)
+
+
+@pytest.mark.parametrize("how", ["outer", "inner", "left", "right"])
+@pytest.mark.parametrize("cat", [True, False])
+def test_join_multi(how, cat):
+
+    size = 10
+    index = ["b", "c"]
+    df1 = pd.DataFrame(
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply python/cudf/cudf/tests/test_joining.py]: I'll try adding tests for nulls --  The PR definitely supports nulls in **non-index** columns, but does not always work for nulls in the MultiIndex.  For string/categorical columns, null index values seems to break the pandas join as well.  For numerical columns, nulls get moved to the end of the column during the `to_frame()` call (causing erroneous behavior).  I will look into a fix for the numerical-column issue.

File: python/cudf/cudf/tests/test_joining.py (hunk 5 above)
Comment (rjzamora) [reply python/cudf/cudf/tests/test_joining.py]: This is from an earlier iteration. We can remove the `sort_values(["a1", "a2"])`

File: python/cudf/cudf/tests/test_joining.py (hunk 3 above)
Comment (rjzamora) [reply python/cudf/cudf/tests/test_joining.py]: Thats right

**Comment by rjzamora** [comment]
@cwharris - Let me know if the tests need more work.

Note that I added a fix for #3137 here, so null values are now tested in numerical multi-index levels.

File: python/cudf/cudf/tests/test_joining.py (hunk 1 above)
Comment (cwharris) [reply python/cudf/cudf/tests/test_joining.py]: I made some erroneous comments. ^^^

`np.arange(size)` doesn't produce random results. My bad.

`np.random.randint(0, 2, size=size)` will likely produce at least one of each value in the range, but it's possible it won't. I'd like to see explicit test data (not random) that will test overlap + lack of overlap on both the left and right sides, and then maybe some random data for good measure.

I see us testing random values against random values in lots of places, and while that's great for testing arithmetic, it's not deterministic when testing discreet value comparisons.

**Comment by rjzamora** [comment]
Oops - Sorry if I merged too soon @kkraus14 ?

**Pull request merged**